lightning/src/ln/channelmanager.rs

```diff
@@ -1437,121 +1447,130 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+			// We have to have some headroom to broadcast on chain if we have the preimage, so make sure we have at least
+			// HTLC_FAIL_BACK_BUFFER blocks to go.
+			// Also, ensure that, in the case of an unknown payment hash, our payment logic has enough time to fail the HTLC backward
+			// before our onchain logic triggers a channel closure (see HTLC_FAIL_BACK_BUFFER rational).
+			if (msg.cltv_expiry as u64) <= self.best_block.read().unwrap().height() as u64 + HTLC_FAIL_BACK_BUFFER as u64 + 1 {
+				return_err!("The final CLTV expiry is too soon to handle", 17, &[0;0]);
+			}
+			// final_incorrect_htlc_amount
+			if next_hop_data.amt_to_forward > msg.amount_msat {
+				return_err!("Upstream node sent less than we were supposed to receive in payment", 19, &byte_utils::be64_to_array(msg.amount_msat));
+			}
+			// final_incorrect_cltv_expiry
+			if next_hop_data.outgoing_cltv_value != msg.cltv_expiry {
+				return_err!("Upstream node set CLTV to the wrong value", 18, &byte_utils::be32_to_array(msg.cltv_expiry));
+			}
 
-				PendingHTLCStatus::Forward(PendingHTLCInfo {
-					routing: PendingHTLCRouting::Receive {
-						payment_data: payment_data.unwrap(),
-						incoming_cltv_expiry: msg.cltv_expiry,
-					},
-					payment_hash: msg.payment_hash.clone(),
-					incoming_shared_secret: shared_secret,
-					amt_to_forward: next_hop_data.amt_to_forward,
-					outgoing_cltv_value: next_hop_data.outgoing_cltv_value,
-				})
-			} else {
-				let mut new_packet_data = [0; 20*65];
-				let read_pos = chacha_stream.read(&mut new_packet_data).unwrap();
-				#[cfg(debug_assertions)]
-				{
-					// Check two things:
-					// a) that the behavior of our stream here will return Ok(0) even if the TLV
-					//    read above emptied out our buffer and the unwrap() wont needlessly panic
-					// b) that we didn't somehow magically end up with extra data.
-					let mut t = [0; 1];
-					debug_assert!(chacha_stream.read(&mut t).unwrap() == 0);
-				}
-				// Once we've emptied the set of bytes our peer gave us, encrypt 0 bytes until we
-				// fill the onion hop data we'll forward to our next-hop peer.
-				chacha_stream.chacha.process_in_place(&mut new_packet_data[read_pos..]);
+			let routing = match next_hop_data.format {
+				msgs::OnionHopDataFormat::Legacy { .. } => return_err!("We require payment_secrets", 0x4000|0x2000|3, &[0;0]),
+				msgs::OnionHopDataFormat::NonFinalNode { .. } => return_err!("Got non final data with an HMAC of 0", 0x4000 | 22, &[0;0]),
+				msgs::OnionHopDataFormat::FinalNode { payment_data, keysend_preimage } => {
+					if payment_data.is_some() && keysend_preimage.is_some() {
+						return_err!("We don't support MPP keysend payments", 0x4000|22, &[0;0]);
+					} else if let Some(data) = payment_data {
+						PendingHTLCRouting::Receive {
+							payment_data: data,
+							incoming_cltv_expiry: msg.cltv_expiry,
+						}
+					} else if let Some(payment_preimage) = keysend_preimage {
+						PendingHTLCRouting::ReceiveKeysend {
```

Comment:
What do you think about checking early that keysend preimage == HTLC hash and isn't pure junk, returning 0x4000|22 if it is ? May help with probing concerns expressed after.

Reply:
Ah! Yea, if we do that then we don't have to completely rewrite this whole PR to avoid the issues you noted at https://github.com/rust-bitcoin/rust-lightning/pull/967#discussion_r670519040

Reply:
Yes i think that's okay to solve the issues. I had a concern with MPP payment where a preimage revealed could be replay as a probing keysend as long as the whole eMPP isn't settled but in fact we won't start the MPP settlement phase until all chunks have been received. And when we do so we remove the entry from `pending_inbound_payments` and as such the collision ?

Reply:
I think it doesn't matter either way - basically we would be using the payment preimage the same way we use the payment secret today - to authenticate that the sender of a given HTLC is the one authorized to send for that payment hash. We reliably check this authentication immediately upon HTLC receipt both for MPP and regular payments, and will just have to also do it for keysend.

Reply:
> What do you think about checking early that keysend preimage == HTLC hash and isn't pure junk, returning 0x4000|22 if it is ? May help with probing concerns expressed after.

Added this check! Thanks for the thoughtful review Antoine